File: integration/todobackend/php/playback_test.php
<?php

declare(strict_types=1);

/**
 * TodoBackend browser integration test — PLAYBACK phase (the CI artifact).
 *
 * Replays the committed CRUD tape through a Servirtium VCR and runs the real
 * TodoBackend Mocha spec against it in headless Chrome (PHP php-webdriver). No
 * SUT, no network — the whole CRUD conversation comes off the tape. This is the
 * offline test wired into aeb (.php_playback.ae); record.php regenerates it.
 *
 * Run via the .php_playback.ae node, or directly:
 *   SERVIRTIUM_VCR_LIB=../../../core/native/libservirtium_vcr.so \
 *   CHROMEDRIVER=<path> php playback_test.php
 */

// The PHP binding's composer autoloader registers both Servirtium\ (from
// php/src) and php-webdriver/webdriver (from php/vendor).
require __DIR__ . '/../../../php/vendor/autoload.php';
require __DIR__ . '/browser.php';

use Servirtium\Vcr;

function main(): int
{
    $vcr = Vcr::playback(TAPE)
        ->staticContent('/suite', SUITE_DIR)
        ->untaped('/favicon.ico')
        ->port(VCR_PORT)
        ->start();
    try {
        [$passes, $failures, $msgs] = run_suite($vcr->baseUrl());
        echo "mocha (playback): {$passes} passed, {$failures} failed\n";
        foreach ($msgs as $m) {
            echo "  FAIL: {$m}\n";
        }
        $ok = $failures === 0 && $passes > 0;
        echo($ok ? "TODOBACKEND_PLAYBACK_OK\n" : "TODOBACKEND_PLAYBACK_FAIL\n");
        return $ok ? 0 : 1;
    } finally {
        $vcr->stop();
    }
}

exit(main());

File: integration/todobackend/dart/playback_test.dart
/// TodoBackend browser integration test — PLAYBACK phase (the CI artifact).
///
/// Replays the committed CRUD tape through a Servirtium VCR and runs the real
/// TodoBackend Mocha spec against it in headless Chrome (Dart `webdriver`). No
/// SUT, no network — the whole CRUD conversation comes off the tape. This is
/// the offline test wired into aeb (.dart_playback.ae); record.dart regenerates
/// the tape.
///
/// Run via the .dart_playback.ae leaf, or directly from this dir:
///   SERVIRTIUM_VCR_LIB=../../../core/native/libservirtium_vcr.so \
///   dart test -j 1
@TestOn('vm')
library;

import 'package:servirtium/servirtium.dart';
import 'package:test/test.dart';

import 'browser.dart';

void main() {
  test('replays the TodoBackend CRUD tape and passes the Mocha suite in Chrome',
      () async {
    final vcr =
        Vcr.playback(tape).staticContent('/suite', suiteDir).untaped('/favicon.ico').port(vcrPort).start();
    try {
      final (passes, failures, msgs) = await runSuite(vcr.baseUrl);
      // ignore: avoid_print
      print('mocha (playback): $passes passed, $failures failed');
      for (final m in msgs) {
        // ignore: avoid_print
        print('  FAIL: $m');
      }
      expect(failures, 0, reason: msgs.join('\n'));
      expect(passes, greaterThan(0));
    } finally {
      vcr.close();
    }
  }, timeout: const Timeout(Duration(minutes: 3)));
}
